From a306c12ff4fad3a548f748988de3de6df0ed0b50 Mon Sep 17 00:00:00 2001
From: allen yan <yanwei@marvell.com>
Date: Tue, 27 Jun 2017 10:25:01 +0800
Subject: [PATCH 1045/1345] fix: clk: a3700: fix the NULL pointer issue

commit  f7c3675b59d69a9a9abc1722614c529fa79ac97c from
https://github.com/MarvellEmbeddedProcessors/linux-marvell.git

- Currently, there is NULL pointer issue about clks pointer when
  run clock driver's probe function. It is because of that the struct
  clk_hw_onecell_data is missing on branch linux-4.4.52-devel and we
  use clk_onecell_data instead of it:
  struct clk_onecell_data {
	struct clk **clks;
	unsigned int clk_num;
  };
  But the pointer clks isn't be assigned a value, it is NULL.

- This patch fixes this issue.

Change-Id: Iaa5acc657f173bd997f687eb52f66b9f0cbb46e9
Signed-off-by: allen yan <yanwei@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/40839
Tested-by: iSoC Platform CI <ykjenk@marvell.com>
Reviewed-by: Victor Gu <xigu@marvell.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/clk/mvebu/armada-37xx-periph.c |    8 ++++++--
 drivers/clk/mvebu/armada-37xx-tbg.c    |    8 +++++---
 2 files changed, 11 insertions(+), 5 deletions(-)

diff --git a/drivers/clk/mvebu/armada-37xx-periph.c b/drivers/clk/mvebu/armada-37xx-periph.c
index 2895ed6..15aac40 100644
--- a/drivers/clk/mvebu/armada-37xx-periph.c
+++ b/drivers/clk/mvebu/armada-37xx-periph.c
@@ -391,12 +391,16 @@ static int armada_3700_periph_clock_probe(struct platform_device *pdev)
 	if (!driver_data)
 		return -ENOMEM;
 
-	driver_data->hw_data = devm_kzalloc(dev, sizeof(*driver_data->hw_data) +
-			    sizeof(*driver_data->hw_data->hws) * num_periph,
+	driver_data->hw_data = devm_kzalloc(dev, sizeof(*driver_data->hw_data),
 			    GFP_KERNEL);
 	if (!driver_data->hw_data)
 		return -ENOMEM;
 	driver_data->hw_data->clk_num = num_periph;
+	driver_data->hw_data->clks = devm_kzalloc(dev, driver_data->hw_data->clk_num *
+						  sizeof(*driver_data->hw_data->clks),
+						  GFP_KERNEL);
+	if (!driver_data->hw_data->clks)
+		return -ENOMEM;
 
 	spin_lock_init(&driver_data->lock);
 
diff --git a/drivers/clk/mvebu/armada-37xx-tbg.c b/drivers/clk/mvebu/armada-37xx-tbg.c
index 21aa133..93a7c3f 100644
--- a/drivers/clk/mvebu/armada-37xx-tbg.c
+++ b/drivers/clk/mvebu/armada-37xx-tbg.c
@@ -91,12 +91,14 @@ static int armada_3700_tbg_clock_probe(struct platform_device *pdev)
 	void __iomem *reg;
 	int i, ret;
 
-	hw_tbg_data = devm_kzalloc(&pdev->dev, sizeof(*hw_tbg_data)
-				   + sizeof(*hw_tbg_data->hws) * NUM_TBG,
-				   GFP_KERNEL);
+	hw_tbg_data = devm_kzalloc(&pdev->dev, sizeof(*hw_tbg_data), GFP_KERNEL);
+
 	if (!hw_tbg_data)
 		return -ENOMEM;
 	hw_tbg_data->clk_num = NUM_TBG;
+
+	hw_tbg_data->clks = devm_kzalloc(&pdev->dev, hw_tbg_data->clk_num * sizeof(*hw_tbg_data->clks),
+				    GFP_KERNEL);
 	platform_set_drvdata(pdev, hw_tbg_data);
 
 	parent = devm_clk_get(dev, NULL);
-- 
1.7.9.5

